src/tokens/fungible.rs

```diff
@@ -0,0 +1,528 @@
+        &mut self,
+        operator: Operator,
+        delta_amount: Amount,
+    ) -> Result<FTApproval, FTError> {
+        let msg_source = msg::source();
+        let amount =
+            self.change_allowance(msg_source, operator, delta_amount, Amount::checked_add)?;
+
+        self.internal_approve(msg_source, operator, amount)
+    }
+
+    /// Decreases by `delta_amount` an allowance for `operator` to spend
+    /// [`msg::source()`]'s tokens.
+    ///
+    /// # Errors
+    /// - [`FTError::InsufficientAllowance`] if `operator` doesn't have given
+    /// `delta_amount` of an allowance.
+    pub fn decrease_allowance(
+        &mut self,
+        operator: Operator,
+        delta_amount: Amount,
+    ) -> Result<FTApproval, FTError> {
+        let msg_source = msg::source();
+        let amount =
+            self.change_allowance(msg_source, operator, delta_amount, Amount::checked_sub)?;
+
+        self.internal_approve(msg_source, operator, amount)
+    }
+
+    /// Mints to `to` `amount` of tokens.
+    ///
+    /// # Errors
+    /// - [`FTError::ZeroRecipientAddress`] if `to` is [`ActorId::zero()`].
+    /// - [`FTError::InsufficientAmount`] if given `amount` with the current
+    /// total token supply overflows the [`Amount`] type.
+    pub fn mint(&mut self, to: ActorId, amount: Amount) -> Result<FTTransfer, FTError> {
+        if to.is_zero() {
+            return Err(FTError::ZeroRecipientAddress);
+        }
+
+        if let Some(total_supply) = self.total_supply.checked_add(amount) {
+            self.total_supply = total_supply;
+        } else {
+            return Err(FTError::InsufficientAmount);
+        }
+
+        self.owners.entry(to).or_default().balance += amount;
+
+        Ok(FTTransfer {
+            from: ActorId::zero(),
+            to,
+            amount,
+        })
+    }
+
+    /// Burns from `from` `amount` of tokens.
+    ///
+    /// # Errors
+    /// - [`FTError::ZeroSenderAddress`] if `from` is [`ActorId::zero()`].
+    /// - [`FTError::InsufficientAmount`] if `from` doesn't have given `amount`
+    /// of tokens.
+    pub fn burn(&mut self, from: Owner, amount: Amount) -> Result<FTTransfer, FTError> {
+        if from.is_zero() {
+            return Err(FTError::ZeroSenderAddress);
+        }
+
+        self.burn_balance(from, amount)?;
+        self.total_supply -= amount;
+
+        Ok(FTTransfer {
+            from,
+            to: ActorId::zero(),
+            amount,
+        })
+    }
+}
+
+/// Fungible token error variants.
+#[derive(Debug, Encode, Decode, PartialEq, Eq, PartialOrd, Ord, Clone, Copy, TypeInfo, Hash)]
+pub enum FTError {
+    /// Token owner doesn't have a sufficient amount of tokens. Or there was the
+    /// [`Amount`] overflow during token minting.
+    InsufficientAmount,
+    /// [`msg::source()`] or operator doesn't have a sufficient allowance of
+    /// tokens. Or there was the [`Amount`] overflow during allowance
+    /// increasing.
+    InsufficientAllowance,
```

Comment:
What's the meaning of the `NotAllowed` variant? The overflow error? I think it's clear that `InsufficientAllowance` returned from allowance changing methods means the overflow error because a sufficient allowance implies amount that doesn't create an overflow.

Reply:
I meant returning `NotAllowed` when no approvals exist for a given account id. But I don't know whether it is the case for library users.

Reply:
Then this is the same case as returning `Option` from `balance_of()`. The PSP standart says that the absence of the allowance equals 0 allowance. And ERC-20 also states the same.